swarm-test/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+    async fn dial_and_wait(&mut self, addr: Multiaddr) -> PeerId;
+
+    /// Wait for specified condition to return `Some`.
+    async fn wait<E, P>(&mut self, predicate: P) -> E
+    where
+        P: Fn(
+            SwarmEvent<<Self::NB as NetworkBehaviour>::OutEvent, THandlerErr<Self::NB>>,
+        ) -> Option<E>,
+        P: Send;
+
+    /// Listens for incoming connections, polling the [`Swarm`] until the transport is ready to accept connections.
+    ///
+    /// The first address is for the memory transport, the second one for the TCP transport.
+    async fn listen(&mut self) -> (Multiaddr, Multiaddr);
+
+    /// Returns the next [`SwarmEvent`] or times out after 10 seconds.
+    ///
+    /// If the 10s timeout does not fit your usecase, please fall back to `StreamExt::next`.
+    async fn next_swarm_event(
+        &mut self,
+    ) -> SwarmEvent<<Self::NB as NetworkBehaviour>::OutEvent, THandlerErr<Self::NB>>;
+
+    /// Returns the next behaviour event or times out after 10 seconds.
+    ///
+    /// If the 10s timeout does not fit your usecase, please fall back to `StreamExt::next`.
+    async fn next_behaviour_event(&mut self) -> <Self::NB as NetworkBehaviour>::OutEvent;
+
+    async fn loop_on_next(self);
+}
+
+/// Drives two [`Swarm`]s until a certain number of events are emitted.
+///
+/// # Usage
+///
+/// ## Number of events
+///
+/// The number of events is configured via const generics based on the array size of the return type.
+/// This allows the compiler to infer how many events you are expecting based on how you use this function.
+/// For example, if you expect the first [`Swarm`] to emit 2 events, you should assign the first variable of the returned tuple value to an array of size 2.
+/// This works especially well if you directly pattern-match on the return value.
+///
+/// ## Type of event
+///
+/// This function utilizes the [`TryIntoOutput`] trait.
+/// Similar as to the number of expected events, the type of event is inferred based on your usage.
+/// If you match against a [`SwarmEvent`], the first [`SwarmEvent`] will be returned.
+/// If you match against your [`NetworkBehaviour::OutEvent`] type, [`SwarmEvent`]s which are not [`SwarmEvent::Behaviour`] will be skipped until the [`Swarm`] returns a behaviour event.
+///
+/// You can implement the [`TryIntoOutput`] for any other type to further customize this behaviour.
+///
+/// # Difference to [`futures::future::join`]
+///
+/// This function is similar to joining two futures with two crucial differences:
+/// 1. As described above, it allows you to obtain more than a single event.
+/// 2. More importantly, it will continue to poll the [`Swarm`]s **even if they already has emitted all expected events**.
+///
+/// Especially (2) is crucial for our usage of this function.
+/// If a [`Swarm`] is not polled, nothing within it makes progress.
+/// This can "starve" the other swarm which for example may wait for another message to be sent on a connection.
```

Comment:
I spent a good two hours of debugging timeouts in the test and unwoken tasks until it dawned on me why things were not working. @elenaf9 I believe this might also be the reason for your timeouts in #3345.

This function may seem like a beast but after trying various solutions locally, I am convinced it makes sense to add this to the test-harness and use this instead of `future::join` for testing `Swarm`s.